client/data/media/use-edit-media-mutation.js

```diff
@@ -0,0 +1,69 @@
+import { createHigherOrderComponent } from '@wordpress/compose';
+import { useTranslate } from 'i18n-calypso';
+import { useCallback } from 'react';
+import { useMutation } from 'react-query';
+import { useDispatch } from 'react-redux';
+import { createTransientMedia } from 'calypso/lib/media/utils';
+import { reduxGetState } from 'calypso/lib/redux-bridge';
+import wp from 'calypso/lib/wp';
+import { editMediaItem, receiveMedia } from 'calypso/state/media/actions';
+import { gutenframeUpdateImageBlocks } from 'calypso/state/media/thunks';
+import { errorNotice, removeNotice } from 'calypso/state/notices/actions';
+import getMediaItem from 'calypso/state/selectors/get-media-item';
+
+export function useEditMediaMutation( queryOptions ) {
+	return useMutation(
+		( { siteId, mediaId, payload } ) => wp.site( siteId ).media( mediaId ).edit( payload ),
+		queryOptions
+	);
+}
+
+export const withEditMedia = createHigherOrderComponent(
+	( Wrapped ) => ( props ) => {
+		const dispatch = useDispatch();
+		const translate = useTranslate();
+		const { mutate } = useEditMediaMutation( {
+			onMutate( { mediaId } ) {
+				dispatch( removeNotice( `update-media-notice-${ mediaId }` ) );
+			},
+			onSuccess( media, { siteId } ) {
+				dispatch( receiveMedia( siteId, media ) );
+				dispatch( gutenframeUpdateImageBlocks( media, 'updated' ) );
+			},
+			onError( error, { siteId, originalMediaItem } ) {
+				dispatch( receiveMedia( siteId, originalMediaItem ) );
+				dispatch(
+					errorNotice( translate( 'We were unable to process this media item.' ), {
+						id: `update-media-notice-${ originalMediaItem.ID }`,
+					} )
+				);
+			},
+		} );
+
+		const editMedia = useCallback(
+			( siteId, item ) => {
+				const transientMediaItem = createTransientMedia( item.media || item.media_url );
+
+				if ( ! transientMediaItem ) {
+					return;
+				}
+
+				const mediaId = item.ID;
+				const originalMediaItem = getMediaItem( reduxGetState(), siteId, mediaId );
```

Comment:
Please note, this usage of `reduxGetState` (from `redux-bridge`) is temporary and will be removed once we migrated data fetching to React Query as well. At that point we can retrieve the `originalMediaItem` from the query cache directly.

I think we could also retrieve all media inside `withEditMedia` with `getMedia` selector to have it in context for `editMedia` but that seemed like overkill. Another way would be to expect the original media item to be passed via `editMedia` (also only temporarily) /cc @jsnajdr if any of that is preferred over usage of `redux-bridge`.

Reply:
You can create a thunk called `startEditMediaItem( siteId, item )` which can call `getState` to get the original item from Redux, call `createTransientMedia` and `dispatch( editMediaItem )`, and finally return the `originalMediaItem`. The `editMedia` callback becomes:
```js
( siteId, item ) => {
  const originalItem = dispatch( startEditMediaItem( siteId, item ) );
  if ( originalItem ) {
    mutate( siteId, item.ID, payload: item, originalItem );
  }
}
```
See how almost everything has been extracted into the thunk?